src/ncp/ncp_host.cpp

```diff
@@ -71,6 +75,50 @@ void NcpHost::GetDeviceRole(DeviceRoleHandler aHandler)
     mNcpSpinel.GetDeviceRole(aHandler);
 }
 
+void NcpHost::Join(const otOperationalDatasetTlvs &aActiveOpDatasetTlvs, const AsyncResultReceiver aReceiver)
+{
+    ChainedAsyncResultPtr result;
+    bool                  isAlready    = !mJoinReceivers.empty();
+    auto                  errorHandler = [this](otError aError, const std::string &aErrorInfo) {
+        CallAsyncReceivers(mJoinReceivers, aError, aErrorInfo);
+    };
+
+    mJoinReceivers.push_back(aReceiver);
+    VerifyOrExit(!isAlready);
+
+    result = std::make_shared<ChainedAsyncResult>(errorHandler);
+    result
+        ->First([this, &aActiveOpDatasetTlvs](ChainedAsyncResultPtr aNext) {
+            mNcpSpinel.DatasetSetActiveTlvs(aActiveOpDatasetTlvs, aNext);
+        })
+        ->Then([this](ChainedAsyncResultPtr aNext) { mNcpSpinel.Ip6SetEnabled(true, aNext); })
+        ->Then([this](ChainedAsyncResultPtr aNext) { mNcpSpinel.ThreadSetEnabled(true, aNext); });
+    result->Run();
```

Comment:
How do you ensure that `result` is still alive when it goes out of this method's scope?

Reply:
So long as we called any NcpSpinel methods that takes this shared pointer. Because in the NcpSpinel methods, the shared pointer will be stored until there is a response. When there is a response, the shared pointer will be reset and the object will be released.